src/Security/Core/OAuthUserProvider.php

```diff
@@ -2,74 +2,43 @@
 
 namespace App\Security\Core;
 
+use App\Entity\User\User;
 use HWI\Bundle\OAuthBundle\OAuth\Response\UserResponseInterface;
-use HWI\Bundle\OAuthBundle\Security\Core\User\FOSUBUserProvider;
-use HWI\Bundle\OAuthBundle\Security\Core\Exception\AccountNotLinkedException;
+use HWI\Bundle\OAuthBundle\Security\Core\User\EntityUserProvider;
+use Symfony\Component\Security\Core\Exception\UsernameNotFoundException;
 
-class OAuthUserProvider extends FOSUBUserProvider
+class OAuthUserProvider extends EntityUserProvider
 {
     /**
      * {@inheritdoc}
      */
     public function loadUserByOAuthUserResponse(UserResponseInterface $response)
     {
-        /**
-         * @var string id in 3rd party system
-         * For facebook this is user id
-         */
-        $idFromThirdParty = $response->getUsername();
-
-        if (empty($idFromThirdParty)) {
-            throw new AccountNotLinkedException(sprintf("User '%s' not found.", $idFromThirdParty));
-        }
-
-        $field = $this->getProperty($response);
-        $user = $this->userManager->findUserBy([$field => $idFromThirdParty]);
-
-        if (!$user) {
-            // if user email is changed in 3rd party system, then change our data too
-            // the problem is when user have login from facebook and google and emails are different :(
-            if ($response->getEmail()) {
-                $user = $this->userManager->findUserByEmail($response->getEmail());
-            }
+        try {
+            return parent::loadUserByOAuthUserResponse($response);
+        } catch (UsernameNotFoundException $e) {
+            $email = $response->getEmail();
+            /**
+             * @var $user User
+             */
+            $user = $this->findUser(['email' => $email]);
 
             if ($user) {
-                $user->setFacebookId($idFromThirdParty);
+                $user->setFacebookId($response->getUsername());
+            } else {
+                $user = new User();
+                $user->setNickname($response->getNickname());
+                $user->setEmail($response->getEmail());
+                $user->setFacebookId($response->getUsername());
+                $user->acceptTerms();
```

Comment:
I'm not sure what is the problem with duplicate usernames (renamed to nickname, e.g. person's name). If you have Ivan Ivanov with facebookid == xxxx1 (username) and Ivan Ivanov with facebookid == xxxx2 then you rename his name. Why since login is done by facebookid?

Reply:
In the future we may have login by "username or email". We will not be able to add such functionality with duplicated usernames.

Reply:
@angelk Which field will be mapped by username? The real name (aka nickname) or the facebookid (aka username)?

Reply:
Facebook id is "random number", so our username is facebook real name

Reply:
Then there should be different fields for real name and nickname. For instance:
- "real_name" or just "name" - non-unique, without canonical representation
- 'nickname' - unique, with canonical. Auto-generated (like in the previous version), but the user should have the ability to change it. Like the github username, e.g. angelk. :)

That sounds OK?

Reply:
I don't think we need "real name" and "nickname", we can use only 1.
We may allow user to change it.
Facebook is linked by id and email; the username is not important.